### ISSUE TITLE
accounts-facebook Error: This authorization code has been used.

### ISSUE THREAD

**Issue by arnonate** [meteor/meteor#3742]
I have a current live Meteor app that leverages accounts-facebook for user accounts. I have an issue where a specific user is not able to sign in. The logs show this error...

{"line":"398","file":"oauth_server.js","message":"Error in OAuth Server: Failed to complete OAuth handshake with Facebook. failed [400] {\"error\":{\"message\":\"This authorization code has been used.\",\"type\":\"OAuthException\",\"code\":100}}","time":{"$date":1424105015737},"level":"warn"}

The following is my accounts creation code...

```
// first, remove configuration entry in case service is already configured
ServiceConfiguration.configurations.remove({
    service: "facebook"
});

ServiceConfiguration.configurations.insert({
    service: "facebook",
    appId: "<id>",
    secret: "<secret>"
});

Accounts.onCreateUser(function(options, user) {
    // We still want the default hook's 'profile' behavior.
    if (options.profile) {
        user.profile = options.profile;
        user.profile.memberSince = new Date();

        // Copy data from Facebook to user object
        user.profile.facebookId = user.services.facebook.id;
        user.profile.firstName = user.services.facebook.first_name;
        user.profile.email = user.services.facebook.email;
        user.profile.link = user.services.facebook.link;
    }

    return user;
});
```

**Comment by avital** [comment]
Does this happen consistently? I've seen this happen before, but wasn't able to reproduce (the subsequent login would work).

(The `Accounts.onCreateUser` call shouldn't be related as that is executed after the login handshake is complete)

**Comment by arnonate** [comment]
This is happening specifically for one user on my site. He cannot log in and every time he tries, that is the error I receive in my logs.

**Comment by Tarang** [comment]
This happens to us. It works when I log in but if I check our server logs over a few weeks there are a few of OAuth login errors in there.

**Comment by glasser** [comment]
http://stackoverflow.com/questions/16224141/can-not-authorise-user-in-facebook-js-sdk-this-authorization-code-has-been-us

**Comment by arnonate** [comment]
@glasser I saw that post... but I don't really understand how I can accomplish that using Meteor's accounts-facebook package.

**Comment by lpender** [comment]
I'm also having this issue and feeling quite stumped.

**Comment by lpender** [comment]
This helped: http://bulenttastan.net/login-with-facebook-using-meteor-js/

**Comment by luccamordente** [comment]
Any updates on this issue?

**Comment by elie222** [comment]
+1

**Comment by juliomac** [comment]
+1

**Comment by aedm** [comment]
+1

**Comment by sean-stanley** [comment]
+1 I cannot login to my own app with facebook. It's possible all users are effected? I can guarantee that my facebook app is live and the credentials are correct. I've tried creating an all new facebook app twice and putting the new credentials in but still no luck.

**Comment by JREAM** [comment]
Facebook Issue here also, same thing

**Comment by weeger** [comment]
+10

**Comment by srikieonline** [comment]
+20

**Comment by ryanmadden** [comment]
+1

**Comment by brianscroggins24** [comment]
Any solutions found? Having a similar issue

**Comment by nilsi** [comment]
Like @ryanmadden says I also suspect this has to do with http after reading this.

http://stackoverflow.com/questions/30690500/possible-reason-for-freezing-meteor-login-with-facebook-on-heroku

I got exact same issue and after looking into my Facebook developer settings for my app I found that I used http for my redirect addresses. I just changed it and hope it will help. Gonna keep track if I get any more of these issues.

**Comment by nilsi** [comment]
It did not solve the problem but I can confirm that it happens when a user accessing the site via http rather than https. Any fix for this?

**Comment by avital** [comment]
You should really not have users logging in to your site via http. Add the
force-ssl package.

**Comment by nilsi** [comment]
@avital yes, I already have force-ssl. Seems like this first time ever a user visits the site over http the redirecting does not work. After refreshing the site it works however and redirect to https. This might be a bug in the force-ssl package then. Im not sure how to tackle this. I tried in another browser and can confirm that this happens but hard to debug.

**Comment by mcoenca** [comment]
Any update on this one ? I've got one user (my client, of course) than can't login to our app because of this. We're deployed on Galaxy, is the force-ssl package still necessary ?

**Comment by hwillson** [comment]
Reading over Facebook's OAuth docs, it looks like a potential solution to this problem involves swapping the current short-lived access token with a long-lived token. Facebook outlines how to do this in the [Expiration and Extension of Access Tokens](https://developers.facebook.com/docs/facebook-login/access-tokens/expiration-and-extension) section of their docs.

I'll mark this a `pull-requests-encouraged`. If anyone is interested in tackling this, take a look at how the short-lived access token is currently fetched [here](https://github.com/meteor/meteor/blob/355cde306a57ec30a13dd5ede24c9936f93ae76b/packages/facebook-oauth/facebook_server.js#L56-L64). Using a short-lived/long-lived token will need to be configurable (with short-lived being the default), and any proposed changes should be backwards compatible. Thanks!

**Comment by elie222** [comment]
is there any sort of workaround to this problem? been open for 2 years and i can see it in our logs today

**Comment by jirikrepl** [comment]
Hi, just reporting, I saw this again today on my localhost env. 

```
{"line":"398","file":"oauth_server.js","message":"Error in OAuth Server: Failed to complete OAuth handshake with Facebook. failed [400] {"error":{"message":"This authorization code has been used.","type":"OAuthException","code":100}}","time":{"$date":1424105015737},"level":"warn"}
```

**Comment by doannucphys** [comment]
I got the same problem with Meteor web, one user can not login with facebook. I verified code but nothing wrong. Event i delete that user in DB and re-insert but nothing change. Then, solution is just clear cache on browser of user (local storage, cookies on his/her chrome browser) and every thing is ok. I use Meteor 1.8x, DB mongodb, client Vuejs.

**Comment by milanzigmond** [comment]
+1 also recently experienced this problem, clients on mobile devices were complaining about it
